src/iam/certhandler/certmodule.cpp

```diff
@@ -0,0 +1,287 @@
+{
+    return mValidCerts;
+}
+
+Error CertModule::SetOwner(const String& password)
+{
+    return mSecureStorage->SetOwner(password);
+}
+
+Error CertModule::Clear()
+{
+    mValidCerts.Clear();
+    return mSecureStorage->Clear();
+}
+
+RetWithError<crypto::PrivateKey*> CertModule::CreateKey(const String& password)
+{
+    auto err = RemoveInvalidCerts(password);
+    if (!err.IsNone()) {
+        return {nullptr, err};
+    }
+
+    err = RemoveInvalidKeys(password);
+    if (!err.IsNone()) {
+        return {nullptr, err};
+    }
+
+    return mSecureStorage->CreateKey(password, mModuleConfig.mKeyGenAlgorithm);
+}
+
+Error CertModule::CreateCSR(const String& subject, const crypto::PrivateKey& privKey, crypto::x509::CSR& csr)
+{
+    crypto::x509::CSR templ;
+    templ.mSubject = subject;
+    templ.mDnsNames = mModuleConfig.mAlternativeNames;
+
+    static constexpr auto oidExtensionExtendedKeyUsage = "2.5.29.37";
+    static constexpr auto oidExtKeyUsageClientAuth = "1.3.6.1.5.5.7.3.1";
+    static constexpr auto oidExtKeyUsageServerAuth = "1.3.6.1.5.5.7.3.2";
+
+    crypto::asn1::ObjectIdentifier oid;
+    for (const auto& extKeyUsage : mModuleConfig.mExtendedKeyUsage) {
+        switch (extKeyUsage.GetValue()) {
+        case ExtendedKeyUsageEnum::eClientAuth:
+            oid = oidExtKeyUsageClientAuth;
+            break;
+
+        case ExtendedKeyUsageEnum::eServerAuth:
+            oid = oidExtKeyUsageServerAuth;
+            break;
+
+        default:
+            LOG_WRN() << "Unexpected extended key usage value: " << extKeyUsage.ToString();
+            break;
+        }
+    }
+
+    if (!oid.IsEmpty()) {
+        StaticArray<uint8_t, crypto::cCertExtValueSize> extValue;
+        crypto::asn1::Encode(oid, extValue);
+
+        crypto::asn1::Extension ext;
+        ext.mId = oidExtensionExtendedKeyUsage;
+        ext.mValue.PushBack(extValue);
+
+        templ.mExtraExtensions.PushBack(ext);
+    }
+
+    return mX509Provider->CreateCSR(templ, privKey, csr);
+}
+
+Error CertModule::ApplyCert(const Array<uint8_t>& pemCert, CertInfo& info)
+{
+    StaticArray<crypto::x509::Certificate, cCertChainSize> certificates;
+    auto                                                   err = mX509Provider->PEMToX509Certs(pemCert, certificates);
+    if (!err.IsNone()) {
+        return err;
+    }
+
+    err = CheckCertChain(certificates);
+    if (!err.IsNone()) {
+        return err;
+    }
+
+    StaticString<cPasswordLen> password;
+    err = mSecureStorage->ApplyCert(certificates, info, password);
+    if (!err.IsNone()) {
+        return err;
+    }
+
+    if (mCertListener != nullptr) {
+        mCertListener->OnCertAdded(GetCertType(), info);
+    }
+
+    return TrimCerts(password);
+}
+
+Error CertModule::CreateSelfSignedCert(const String& password)
+{
+    auto key = CreateKey(password);
+    if (!key.mError.IsNone()) {
+        return key.mError;
+    }
+
+    static constexpr auto cValidSelfSignedCertPeriod = 100; // in years
```

Comment:
Move to private block of cert module. I prefer all consts in one place. As we introduce `Time` module. It should have unit constant inside: such a Time::Hour, Time::Minute, etc. So, value of this const should lool like:
cValidSelfSignedCertPeriod = 100 * 365 * 24 * Time::Hour. We can't have const like Time::Year unfortunately.